xhydro/modelling/obj_funcs.py

```diff
@@ -0,0 +1,751 @@
+
+    epsilon: scalar float indicating the perturbation to add to the flow
+        time series during a transformation to avoid division by zero and
+        logarithmic transformation. The perturbation is equal to:
+
+            perturbation = epsilon * mean(Qobs)
+
+        The default value is 0.01.
+
+
+    NOTE: All data corresponding to NaN values in the observation set are
+          removed from the calculation
+
+    Returns
+    -------
+    obj_fun_val: scalar float representing the value of the selected objective
+    function (obj_fun).
+    """
+    # List of available objective functions
+    obj_func_dict = {
+        "abs_bias": abs_bias,
+        "abs_pbias": abs_pbias,
+        "abs_volume_error": abs_volume_error,
+        "agreement_index": agreement_index,
+        "bias": bias,
+        "correlation_coeff": correlation_coeff,
+        "kge": kge,
+        "kge_mod": kge_mod,
+        "mae": mae,
+        "mare": mare,
+        "mse": mse,
+        "nse": nse,
+        "pbias": pbias,
+        "r2": r2,
+        "rmse": rmse,
+        "rrmse": rrmse,
+        "rsr": rsr,
+        "volume_error": volume_error,
+    }
+
+    # Basic error checking
+    if Qobs.shape[0] != Qsim.shape[0]:
+        sys.exit("Observed and Simulated flows are not of the same size.")
+
+    # Check mask size and contents
+    if mask is not None:
+        # Size
+        if Qobs.shape[0] != mask.shape[0]:
+            sys.exit("Mask is not of the same size as the streamflow data.")
+
+        # All zero or one?
+        if not np.setdiff1d(np.unique(mask), np.array([0, 1])).size == 0:
+            sys.exit("Mask contains values other 0 or 1. Please modify.")
+
+    # Check that the objective function is in the list of available methods
+    if obj_func not in obj_func_dict:
+        sys.exit(
+            "Selected objective function is currently unavailable. "
+            + "Consider contributing to our project at: "
+            + "github.com/hydrologie/xhydro"
+        )
+
+    # Ensure there are no NaNs in the dataset
+    if mask is not None:
+        mask = mask[~np.isnan(Qobs)]
+    Qsim = Qsim[~np.isnan(Qobs)]
+    Qobs = Qobs[~np.isnan(Qobs)]
+
+    # Apply mask before trasform
+    if mask is not None:
+        Qsim = Qsim[mask == 1]
+        Qobs = Qobs[mask == 1]
+        mask = mask[mask == 1]
+
+    # Transform data if needed
+    if transform is not None:
+        Qsim, Qobs = transform_flows(Qsim, Qobs, transform, epsilon)
+
+    # Compute objective function by switching to the correct algorithm. Ensure
+    # that the function name is the same as the obj_func tag or this will fail.
+    function_call = obj_func_dict[obj_func]
+    obj_fun_val = function_call(Qsim, Qobs)
+
+    # Take the negative value of the Objective Function to return to the
+    # optimizer.
+    if take_negative:
+        obj_fun_val = obj_fun_val * -1
+
+    return obj_fun_val
+
+
+def get_objfun_minimize_or_maximize(obj_func):
```

Comment:
```suggestion
def _get_objfun_minimize_or_maximize(obj_func):
```
Les fonctions internes peuvent être cachées en ajoutant `_` au début.

Reply:
Je ne connaissais pas cette possibilité. Cependant, quand je teste avec le underscore, pytest fail car je les importe ailleurs. Devrais-je les cacher si elles sont utilisées ailleurs? Devrais-je modifier les import dans "calibration" pour ajouter le underscore aussi? Je vais le laisser tel quel pour l'instant.